File: solutions/go/matching-brackets/1/matching_brackets.go
package brackets

func Bracket(input string) bool {
	stack := []rune{}
    pairs := map[rune]rune {
        ']': '[',
        '}': '{',
        ')': '(',
    }

    for _, ch := range input {
        switch ch {
            case '[', '{', '(':
            	stack = append(stack, ch)
            case ']', '}', ')':
            	if len(stack) == 0 {
                    return false
                }
                top := stack[len(stack) - 1]
            	if pairs[ch] != top {
                    return false
                }
            	stack = stack[:len(stack) - 1]
        }
    }
    return len(stack) == 0
}

File: solutions/c/matching-brackets/1/matching_brackets.c
#include "matching_brackets.h"
#include <stdlib.h>
#include <string.h>

bool is_paired(const char *input) {
    size_t len = strlen(input);
    char *stack = malloc(sizeof(char) * len);
    size_t j = 0;
    char top;
    for (size_t i = 0; i < len; ++i) {
        switch(input[i]) {
            case '[':
            case '{':
            case '(': 
                stack[j++] = input[i];
                break;
                
            case ']' :
                    if (j == 0) {
                        free(stack);
                        return false;
                    }
                    top = stack[j - 1];
                    if (top != '[') {
                        free(stack);
                        return false;
                    }
                    --j;
                    break;
            case '}' :
                    if (j == 0) {
                        free(stack);
                        return false;
                    }
                    top = stack[j - 1];
                    if (top != '{') {
                        free(stack);
                        return false;
                    }
                    --j;
                    break;
            case ')' :
                    if (j == 0) {
                        free(stack);
                        return false;
                    }
                    top = stack[j - 1];
                    if (top != '(') {
                        free(stack);
                        return false;
                    }
                    --j;
                    break;
        }
    }
    free(stack);
    return j == 0;
}